### ISSUE TITLE
Bugs in `Ensemble.split`

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openpathsampling#437]
## Bug 1: Unexpected behavior of overlap in `split`

Using the notation `A` to mean a frame in state A, and `X` to mean a frame not in state A, what is the expected behavior of:

```python
ensembleAXA.split(trajectoryAXAXA, overlap=1)
```

I would expect this to give 2 trajectories: frames 0,1,2 and frames 2,3,4. It does not. It only gives 0,1,2.

Minimal test:

```python
import openpathsampling as paths
from openpathsampling.tests.test_helpers import make_1d_traj

def xval(snapshot):
    return snapshot.xyz[0][0]

opX = paths.CV_Function(name="opX", f=xval)

stateA = paths.CVRangeVolume(opX, float("-inf"), 0.0)

ensembleAXA = paths.SequentialEnsemble([
    paths.AllInXEnsemble(stateA),
    paths.AllOutXEnsemble(stateA),
    paths.AllInXEnsemble(stateA)
])

trajectoryAXAXA = make_1d_traj(coordinates=[-0.5, 0.5, -0.5, 0.5, -0.5])

ensembleAXA.split(trajectoryAXAXA, overlap=1)
```

Only returns one trajectory. Correct results are obtained with `ensembleAXA.split(trajectoryAXAXA, overlap=2)`.

I think I see how to fix this one, and can do so as part of #435 if you want. I think the problem relates to the question of overshooting vs. not overshooting I mention in #95 (and which is at the heart of "pred", "over", and "hyb" suffixes in my stuff on path ensemble theory.

***

## Bug 2: `Ensemble.split(lazy=False)` errors:

```python
ensembleAXA.split(trajectoryAXAXA, overlap=1, lazy=False)
```

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-2f95ecd8b92c> in <module>()
----> 1 ensembleAXA.split(trajectoryAXAXA, overlap=1, lazy=False)

/Users/dwhs/Dropbox/msm-tis/openpathsampling/ensemble.pyc in split(self, trajectory, lazy, max_length, min_length, overlap)
    432             # overloading (which is not used for Snapshots) and type()
    433             indices = self.find_valid_slices(trajectory.as_proxies(), lazy, max_length,
--> 434                                              min_length, overlap)
    435 
    436             return [paths.Trajectory(trajectory[part]) for part in indices]

/Users/dwhs/Dropbox/msm-tis/openpathsampling/ensemble.pyc in find_valid_slices(self, trajectory, lazy, max_length, min_length, overlap)
    322 #                    print start, start+l
    323                     # test using lazy=False
--> 324                     if self(tt, lazy=False):
    325                         list_left = []
    326                         list_right = []

TypeError: __call__() got an unexpected keyword argument 'lazy'
```

Since I'm not entirely sure what the difference in behavior is supposed to be between `lazy=True` and `lazy=False`, I don't really want to mess with this one. (Do these actually give different results? If not, why do we keep the slower algorithm around?)

**Comment by jhprinz** [comment]
First can easily be fixed 
```py
                    elif self(tt[0:len(tt)-1], trusted=False):
                        ensemble_list.append(slice(start,end-1))
                        pad = min(overlap + 1, end - start - 2)
                        start = end - pad
```

There needs to be a `+ 1` after the overlap. Then it works as expected.

The second can be removed I think. It had to do with the question if `can_append`, `can_prepend` work reliably and will only report `True` if it really makes sense to continue. Anyway, the non-lazy version is so slow that it is not feasible to use, so we might remove it altogether.

I can just add this to the engine module stuff which is almost ready. If you want to work more on #434 and need the split stuff quick then just make a new PR.

**Comment by dwhswenson** [comment]
> There needs to be a + 1 after the overlap. Then it works as expected.

No. Consider the same setup as above, but use

```python
ens3 = paths.LengthEnsemble(3)
ens3.split(trajectoryAXAXA, overlap=1)
```

Current version gives the correct results (same as desired from the `ensembleAXA` above); I'm pretty sure your proposed change would not.

The right way to handle all of this is to follow the lead of `SequentialEnsemble`, where the same problem occurs. The details of why it occurs are discussed a little in #95, but much more in `pathsampling_manuscripts/path_ensemble_theory/path_ensemble_theory.tex` in the subsection "Can-append and can-prepend." `LengthEnsemble`s are perfect-predictive, `AllInXEnsemble`s and `AllOutXEnsemble`s are always-overshoot.

Please let me implement the fix as part of #435, as I said above. This is absolutely required for stuff in #435, and I'll fix it and add appropriate nosetests. I just needed to make sure you hadn't intended different behavior than what I was expecting.

> Anyway, the non-lazy version is so slow that it is not feasible to use, so we might remove it altogether.

I will remove it with my fix in #435.

**Comment by jhprinz** [comment]
Actually, I just tested and it does give the right result as before. I am not sure if the current implementation is close to be efficient, but the missing `+ 1` (only in the second case where  was indeed an error).

So, the simple idea is to 
1. start at the beginning and move the end point to the minimum length
2. continue growing a chunk (add end + 1) with same starting point until `.can_append` returns false. 
3. test if either

    1. the proposed chunk is in the ensemble.If so, add it and move the start to the left of the end of the chunk minus the overlap. Correct this point only to make sure we really move the starting point at least by one.
    2. the proposed chunk without the last element is in the ensemble. If so, add the truncated and move to the end of the truncated (this was the errer) minus overlap and check that we will at least move the starting point by one
    3. Otherwise move the starting point + 1 and set the end to the start + minimum length as in 1.

4. stop if start is so close to end that no more trajectories on min length can be found

I thought that we might improve all this searching a lot, when we add a `min_length`, `max_length` property to ensembles. With
```
    min_length : int
        smallest possible length a trajectory in this ensemble must have, len(traj) >= min_length
    max_length : int
        largest possible length a trajectory in this ensemble can have, len(traj) <= max_length
```

this could be used to remove lots of unnecessary tests and it should be easy to implement.

**Comment by dwhswenson** [comment]
> Actually, I just tested and it does give the right result as before.

Did you check that it gave the *correct* frames? My reading is that it would have given a result, but you would have gotten frames `[0,1,2]` and `[1,2,3]`, instead of `[0,1,2]` and `[2,3,4]`.

> So, the simple idea is to

Something like that. See `SequentialEnsemble` and #95 and the stuff about trimming in `path_ensemble_theory.tex`. I phrase it this way:

1. Run can-append until it fails; this is a candidate trajectory
2. If the candidate trajectory is not in the ensemble, chop off the last frame and try that as the candidate. If still not in the ensemble, then `start` moves by one frame.
3. If the candidate trajectory (possibly after chopping the last frame) is in the ensemble, then you set the new starting point. This works under the perfect-prediction model (again, read `path_ensemble_theory.tex` to see what I mean and why that's a better approach to use.)

> I thought that we might improve all this searching a lot, when we add a min_length, max_length property to ensembles

***Please do not do this***. This makes the mathematical description EXTREMELY difficult. If we stick close to a real mathematical object, proper behavior is guaranteed by mathematics.

**Comment by jhprinz** [comment]
Sorry, forgot to post the result

```
# changed values a little to see frames
trajectoryAXAXA = make_1d_traj(coordinates=[-0.5, 0.4, -0.3, 0.2, -0.1]) 
```
this is copied from the notebook
```
split1 = ensembleAXA.split(trajectoryAXAXA, overlap=0)
for traj in split1:
    print traj.coordinates[:,0,0]

>>> [-0.5  0.4 -0.3]

split2 = ensembleAXA.split(trajectoryAXAXA, overlap=1)
​for traj in split2:
    print traj.coordinates[:,0,0]

>>> [-0.5  0.4 -0.3]
>>> [-0.3  0.2 -0.1]

split3 = ensembleLen3.split(trajectoryAXAXA, overlap=0)
​for traj in split3:
    print traj.coordinates[:,0,0]

>>> [-0.5  0.4 -0.3]

split4 = ensembleLen3.split(trajectoryAXAXA, overlap=1)
​for traj in split4:
    print traj.coordinates[:,0,0]

>>> [-0.5  0.4 -0.3]
>>> [-0.3  0.2 -0.1]
```

> Please do not do this. This makes the mathematical description EXTREMELY difficult. If we stick close to a real mathematical object, proper behavior is guaranteed by mathematics.

??? Sorry, but I do not see why a drived property from a mathematical object makes the description of the object difficult. It is like asking the mathematical question. Does there exist an upper or lower bound on the length of paths that are in this ensemble. This does not have to be exact. For `Length` this can be given exactly. For `Sequence` this is just the sum, `And` and `Or` will translate to using `min`, `max` and so on. It would seem that this can very nicely be translated into a property on the space of ensembles. 

Maybe I wasn't clear: This is just a computed value and a (potentially bad) bound to help with performance, nothing else. It should be easy to implement and improve the speed in splitting.

**Comment by jhprinz** [comment]
There is one point I have not completely confirmed: Are the only chances that `can_append` stops because the current trajectory or with the last frame removed can be a candidate?!? Or are there cases where you would have to cut more from the trajectory. I think you could define ensembles that require this, e.g "fifths last frame should be in A" requires to have an overhead of 5 frames. We do not allow these type of definition, right? And refer to the use of Sequences like `Sequence([InA, Length(4)])`.

**Comment by dwhswenson** [comment]
> Sorry, forgot to post the result

Looks good, then. I misread the code.

> This is just a computed value and a (potentially bad) bound to help with performance, nothing else.

I was understanding a user-defined attribute for all ensembles. That would be a nightmare that would allow users to define things such as `paths.AllInXEnsemble(volume1, min_length=10) & paths.PartOutXEnsemble(volume2, max_length=5)` which has no sense at all (did they mean they want at least 10 frames in `volume1` and at most 5 frames outside of `volume2` -- probably, but that's not what they'd get.) Setting desired lengths is correctly done by using length ensembles.

I still don't see the point to doing this. Again, the approaches based on trimming and such are already well-defined mathematically, and are also the way we've implemented things when they needed to be implemented. The only speed-up you get is that you abort the analysis of the last few frames. Adding more code to save a couple iterations will probably be as/more expensive as just doing iterations. It all sounds like premature optimization.

> There is one point I have not completely confirmed: Are the only chances that can_append stops because the current trajectory or with the last frame removed can be a candidate?!?

Yes. Read `path_ensemble_theory.tex`. This is the poisoned frame theorem (under hybrid can-append). Seriously, read over that stuff, because I've worked out all the mathematics of this in great detail.

> I think you could define ensembles that require this, e.g "fifths last frame should be in A" requires to have an overhead of 5 frames. We do not allow these type of definition, right? And refer to the use of Sequences like Sequence([InA, Length(4)]).

I don't understand what you're saying here. Are you looking for: 

```
SequentialEnsemble([
    InXEnsemble(A) & LengthEnsemble(1), 
    LengthEnsemble(4)
])
```

That's a valid ensemble.

**Comment by jhprinz** [comment]
> I was understanding a user-defined attribute for all ensembles. That would be a nightmare

Agreed, that would be stupid.

> I still don't see the point to doing this.

Yes, not sure. I kind of liked the mathematical correspondence between functions on integers and ensemble types. So I was curious if that could be done elegantly. It could be part of all the mathematical properties of ensembles. I don't think the overhead would really matter. 
The initial idea was sparked by that I realized that the splitting in `FullBootstrap` can take very long. 
Bounds on lengths could (maybe) help here a lot.

But as I am thinking of this. It could be that we should maybe change the splitting strategy. Currently we start at the beginning and search for allowed subparts. My feeling is that we do splitting mostly in two cases. Either 

1. the parts are distributed evenly/randomly in the trajectory or 
2. we have the part we look for at the end. 
3. we usually do not have the case where the searched parts are only at the beginning. If so we could have stopped already. Unless we want some sort of correlation between beginning and end.

So I would propose just to start at the end. This should be faster or equally fast for most cases.

> Seriously, read over that stuff.

I did - a while ago. :)

**Comment by dwhswenson** [comment]
> It could be that we should maybe change the splitting strategy.

I like the idea, but do it as a supplement, not a change.

What you actually want is probably a "last_subtrajectory" function, that returns the last segment satisfying the ensemble. Do this as a separate function.

Reasons not to change the current `split`:

1. Splitting from the end will get you the first example faster, but you still have to do the rest of the trajectory to find all others. Still as slow.
2. The average person would expect `split` to start from t=0. You can make an option to go backward, but that should not be the default.
3. You've mainly seen `split` used for finding a subtrajectory that satisfied a given ensemble, regardless of initial condition. That doesn't mean that's the main way it will be used. I see general analysis as more important in the long run, and there the average user's expectation of behavior is more important.
4. What you actually want isn't about `split` in general, but a specific usage of it. Better to write a custom function for that specific usage.

**Comment by jhprinz** [comment]
Good point. We should have find first and find last in case you need only one. We can (in addition) add optional parameters that set the order in traversing for split and limit the number of found results. 

```
    reverse : bool
        if `True` will start splitting from the end of the trajectory
    n_results : int
        if not zero will stop splitting after at most n_results subtrajectories have been found.
```

Then `find_first_subtrajectory` and `find_last_subtrajectory` can be short-cuts.

**Comment by dwhswenson** [comment]
That sounds good to me.

**Comment by dwhswenson** [comment]
Point of clarification: new version of `split`, with those optional parameters, is to come in a separate PR, right? (I shouldn't be waiting on that in #434?)

**Comment by jhprinz** [comment]
Yes, I will make a separate PR for this.

**Comment by dwhswenson** [comment]
Closed after #441.
